crates/dht/src/hash.rs

```diff
@@ -0,0 +1,443 @@
+//! partitions to create. A factor of 32 will create a single partition. Each unit decrease from 32
+//! will double the number of partitions. A good default for a network that is expected to grow to
+//! a few thousand nodes is 21. This will create 2^11 partitions, which is 2048. Given that data is
+//! also stored redundantly, say 10 times, this means that 20,000 nodes are needed for the data
+//! to be maximally distributed. With higher replication, say 50 times, and a hash factor of 20,
+//! over 200,000 nodes are needed for maximal distribution.
+//!
+//! Each hash partition manages a [PartitionedTime] structure that is responsible for managing the
+//! time slices for that hash partition. The interface of this module is largely responsible for
+//! delegating the updating of time slices to the inner time partitions. This ensures that all the
+//! time partitions are updated in lockstep. This makes reasoning about the space-time partitioning
+//! easier.
+//!
+//! This module must be informed about ops that have been stored. There is no active process here
+//! that can look for newly stored ops. When a batch of ops is stored, the [PartitionedHashes] must
+//! be informed and will split the ops into the right hash partition based on the location of the op.
+//! Ops are then pushed to the inner time partition for each hash partition. It is the time
+//! partitions that are responsible for updating the combined hash values.
+
+use crate::PartitionedTime;
+use kitsune2_api::{
+    ArcLiteral, DynOpStore, K2Error, K2Result, StoredOp, Timestamp,
+};
+use std::collections::HashMap;
+
+/// A partitioned hash structure.
+///
+/// Partitions the hash structure into a configurable number of partitions. Each partition is
+/// responsible for managing the time slices for that partition.
+#[derive(Debug)]
+pub struct PartitionedHashes {
+    size: u32,
+    partitioned_hashes: Vec<HashPartition>,
+}
+
+#[derive(Debug)]
+struct HashPartition {
+    _arc: ArcLiteral,
+    partitioned_time: PartitionedTime,
+}
+
+impl PartitionedHashes {
+    /// Create a new partitioned hash structure.
+    ///
+    /// The `hash_factor` determines the number of partitions to create. A value of 32 will create
+    /// a single partition. Each unit decrease from 32 will double the number of partitions.
+    ///
+    /// Each hash partition owns a [PartitionedTime] structure that is responsible for managing
+    /// the time slices for that hash partition. Other parameters to this function are used to
+    /// create the [PartitionedTime] structure.
+    pub async fn try_from_store(
+        hash_factor: u8,
+        time_factor: u8,
+        current_time: Timestamp,
+        store: DynOpStore,
+    ) -> K2Result<Self> {
+        if hash_factor > 32 {
+            return Err(K2Error::other("Hash factor must be 32 or lower"));
+        }
+
+        let (size, num_buckets) = if hash_factor == 32 {
```

Comment:
```suggestion
        let (size, num_partitions) = if hash_factor == 32 {
```

Reply:
Needs renaming in a few places, done without committing this suggestion :)